File: src/csc_solpacklu.f
!                   ************************
                    SUBROUTINE CSC_SOLPACKLU
!                   ************************
     & (N,M,NZ,LUV,LUR,LUC,VB,VX)
!
!***********************************************************************
! CSC PACKAGE - SERGIO CASTIBLANCO
!***********************************************************************
!
!brief    1) THIS SUBROUTINE SOLVES THE SYSTEM Ax=b WHERE MATRIX A IS LU
!            DECOMPOSITION, i.e., A=LU AND STORED IN CSC-PACKED STORAGE 
!
!history  Sergio Castiblanco
!+        16/02/2022
!+        Translation for original Matlab implementation
!
!-----------------------------------------------------------------------
! CSC-PACKED STORAGE IS THE NEXT:
! As L is Unitary Lower Triangular, i.e., its diagonal are all ones,
! there is no need to store its diagonal, insteadi, in the upper part
! and diagonal the components of matrix U can be stored. Using just one
! matrix to store both L and U. This is done in CSC storage, and it is
! useful for preconditioners as SSOR or ILU, as those have same
! structure as original matrix.
!
!               solve -> Ly = b (forward substitution)
!               solve -> Ux = y (backward substitution)
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!| N        |-->| ROWS OF MATRIX LU                                    |
!| M        |-->| COLUMNS OF MATRIX LU                                 |
!| NZ       |-->| NUMBERS OF NONZERO ELEMNTS OF MATRIX LU              |
!| LUV      |-->| VALUES OF MATRIX LU IN CSC-PACKED STORAGE            |
!| LUR      |-->| ROW POINTERS OF MATRIX LU IN CSC-PACKED STORAGE      |
!| LUC      |-->| COLUMN POINTERS OF MATRIX LU IN CSC-PACKED STORAGE   |
!| VB       |-->| RIGHT HAND SIDE VECTOR b                             |
!| VX       |<->| SOLUTION VECTOR FOR THE SYSTEM                       |
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      USE CSC_STORAGE, EX_CSC_SOLPACKLU => CSC_SOLPACKLU
      IMPLICIT NONE
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      INTEGER, INTENT(IN) :: N,M,NZ
      DOUBLE PRECISION, INTENT(IN), DIMENSION(NZ) :: LUV,LUR
      DOUBLE PRECISION, INTENT(IN), DIMENSION(M+1) :: LUC
      DOUBLE PRECISION, INTENT(IN), DIMENSION(N) :: VB
      DOUBLE PRECISION, INTENT(INOUT), DIMENSION(N) :: VX
!
!  IN SUBROUTINE VARIABLES
!
      INTEGER :: J, I
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
! REPLACING VALUES OF VX WITH THOSE OF VB
!
      DO I=1,N
        VX(I)=VB(I)
      ENDDO
!
! FORWARD SUBSTITUTION
!
      DO J=1,M
        DO I=LUC(J),(LUC(J+1)-LUC(1))
          IF(J.LT.LUR(I))THEN
            VX(LUR(I)) = VX(LUR(I)) - LUV(I)*VX(J)
          ENDDO
        ENDDO
      ENDDO
!
! BACKWARD SUBSTITUTION
!
      DO J=M,1,-1
        DO I = (LUC(J+1)-LUC(1)),LUC(J),-1
          IF(J.EQ.LUR(I))THEN
            VX(J) = VX(J)/LUV(I)
          ELSE IF(J.GT.LUR(I))THEN
            VX(LUR(I)) = VX(LUR(I)) - LUV(I)*VX(J)
          ENDIF
        ENDDO
      ENDDO
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      END SUBROUTINE CSC_SOLPACKLU

File: matlab/csc_solpacklu.m
function x = csc_solpacklu(LUv, LUr, LUc, b)
% This function compute the forward and backward sustitution for the system
% LUx = b, where LU should be entered in pack_csc notation, that's it:
% packLU in Compressed Sparse Column notation.
% Enries are:
%       LUv = vector with values of packLU matrix in CSC format
%       LUr = vector with row indexes of packLU matrix in CSC format
%       LUc = vector with column indexes of packLU matrix in CSC format

x = b;
m = length(LUc)-LUc(1);

%forward sustitution
for j=1:m
    for i = LUc(j):LUc(j+1)-LUc(1)
        if j<LUr(i)
            x(LUr(i)) = x(LUr(i)) - LUv(i)*x(j);
        end
    end
end

%backward sustitution
for j=m:-1:1
    for i = LUc(j+1)-LUc(1):-1:LUc(j)
        if j==LUr(i)
            x(j) = x(j)/LUv(i);
        end
        if j>LUr(i)
            %y(LUr(i)) = y(LUr(i)) - LUv(i)*y(j);
            x(LUr(i)) = x(LUr(i)) - LUv(i)*x(j);
            %LUv(i)
        end
    end
end

end